src/basic/fd-util.h

```diff
@@ -126,3 +127,12 @@ static inline char *format_proc_fd_path(char buf[static PROC_FD_PATH_MAX], int f
 
 #define FORMAT_PROC_FD_PATH(fd) \
         format_proc_fd_path((char[PROC_FD_PATH_MAX]) {}, (fd))
+
+static inline const char *format_fd_path(char buf[static PATH_MAX], int fd) {
+        buf[0] = 0;
+        (void) readlink(FORMAT_PROC_FD_PATH(fd), buf, PATH_MAX);
+        return buf;
+}
+
+#define FORMAT_FD_PATH(fd) \
+        format_fd_path((char[PATH_MAX]) {}, (fd))
```

Comment:
that's 4K on the stack btw. quite a bit...

Reply:
What about using a single static thread local 4k buffer on the stack in the function itself? You'll always override the previous path that way but for this I don't think we care

Reply:
that'd be 4K per thread ahead of time, i doubt that would be better, probably worse. dunno.